File: too-many-cooks/test/agent_streaming_e2e_test.dart
/// E2E agent streaming test — spawn MCP server, open SSE stream
/// on /mcp for an AGENT session, trigger state changes from
/// another agent, ASSERT that notifications arrive over the
/// agent's SSE stream.
///
/// This PROVES that agents receive streamed notifications over
/// their Streamable HTTP SSE connection.
library;

import 'dart:async';
import 'dart:convert';
import 'dart:js_interop';
import 'dart:js_interop_unsafe';

import 'package:dart_node_core/dart_node_core.dart';
import 'package:test/test.dart';
import 'package:too_many_cooks/too_many_cooks.dart' show serverBinary;

// ============================================================
// Named Constants
// ============================================================

const _baseUrl = 'http://localhost:4040';
const _mcpPath = '/mcp';
const _accept = 'application/json, text/event-stream';
const _mcpProtocolVersion = '2025-03-26';
const _defaultEventTimeoutMs = 3000;
const _streamEstablishDelayMs = 200;
const _serverPollDelayMs = 200;
const _eventPollDelayMs = 50;
const _maxServerPollAttempts = 30;

const _agent1Name = 'agent-stream-1';
const _agent2Name = 'agent-stream-2';
const _agent3Name = 'agent-stream-3';

const _testFilePath = '/agent-stream/test.dart';
const _testLockReason = 'agent streaming e2e';
const _testMessageContent = 'hello from agent1';
const _testGoal = 'Agent streaming goal';
const _testTask = 'Testing agent streaming';

const _eventAgentRegistered = 'agent_registered';
const _eventLockAcquired = 'lock_acquired';
const _eventLockReleased = 'lock_released';
const _eventMessageSent = 'message_sent';
const _eventPlanUpdated = 'plan_updated';

const _jsonRpcVersion = '2.0';
const _notificationMethod = 'notifications/message';
const _levelInfo = 'info';

const _dbDir = '.too_many_cooks';
const _dbFiles = ['data.db', 'data.db-wal', 'data.db-shm'];

void main() {
  // Server process shared across all tests
  // ignore: no_late
  late JSObject serverProcess;

  setUpAll(() async {
    _deleteDbFiles();
    serverProcess = _spawnServer();
    await _waitForServer();
  });

  tearDownAll(() {
    _killProcess(serverProcess);
    _deleteDbFiles();
  });

  test('agent receives message_sent notification via SSE', () async {
    await _resetServer();
    final agent1 = _McpClient();
    final agent2 = _McpClient();
    await agent1.initSession();
    await agent2.initSession();

    final reg1 = _parseJson(
      await agent1.callTool('register', {'name': _agent1Name}),
    );
    final reg2 = _parseJson(
      await agent2.callTool('register', {'name': _agent2Name}),
    );
    final key1 = reg1['agent_key']! as String;

    final sse = await _AgentSseClient.connect(agent2.sessionId);

    await agent1.callTool('message', {
      'action': 'send',
      'agent_key': key1,
      'to_agent': reg2['agent_name']! as String,
      'content': _testMessageContent,
    });

    final events = await sse.waitForEvents(1);
    sse.close();

    expect(
      events.isNotEmpty,
      isTrue,
      reason: 'Agent2 MUST receive SSE event for message',
    );
    final eventType = _extractEventType(events.first);
    expect(
      eventType,
      equals(_eventMessageSent),
      reason: 'Event type MUST be $_eventMessageSent',
    );
  });

  test('agent receives lock_acquired notification via SSE', () async {
    await _resetServer();
    final agent1 = _McpClient();
    final agent2 = _McpClient();
    await agent1.initSession();
    await agent2.initSession();

    final reg1 = _parseJson(
      await agent1.callTool('register', {'name': _agent1Name}),
    );
    await agent2.callTool('register', {'name': _agent2Name});
    final key1 = reg1['agent_key']! as String;

    final sse = await _AgentSseClient.connect(agent2.sessionId);

    await agent1.callTool('lock', {
      'action': 'acquire',
      'file_path': _testFilePath,
      'agent_key': key1,
      'reason': _testLockReason,
    });

    final events = await sse.waitForEvents(1);
    sse.close();

    expect(
      events.isNotEmpty,
      isTrue,
      reason: 'Agent2 MUST receive SSE event for lock acquire',
    );
    final eventType = _extractEventType(events.first);
    expect(
      eventType,
      equals(_eventLockAcquired),
      reason: 'Event type MUST be $_eventLockAcquired',
    );
  });

  test('agent receives agent_registered notification via SSE', () async {
    await _resetServer();
    final agent1 = _McpClient();
    final agent2 = _McpClient();
    await agent1.initSession();
    await agent2.initSession();

    await agent1.callTool('register', {'name': _agent1Name});
    await agent2.callTool('register', {'name': _agent2Name});

    final sse = await _AgentSseClient.connect(agent2.sessionId);

    // Register a third agent — agent2 should get notified
    await agent1.callTool('register', {'name': _agent3Name});

    final events = await sse.waitForEvents(1);
    sse.close();

    expect(
      events.isNotEmpty,
      isTrue,
      reason: 'Agent2 MUST receive SSE event for new registration',
    );
    final eventType = _extractEventType(events.first);
    expect(
      eventType,
      equals(_eventAgentRegistered),
      reason: 'Event type MUST be $_eventAgentRegistered',
    );
  });

  test('agent receives plan_updated notification via SSE', () async {
    await _resetServer();
    final agent1 = _McpClient();
    final agent2 = _McpClient();
    await agent1.initSession();
    await agent2.initSession();

    final reg1 = _parseJson(
      await agent1.callTool('register', {'name': _agent1Name}),
    );
    await agent2.callTool('register', {'name': _agent2Name});
    final key1 = reg1['agent_key']! as String;

    final sse = await _AgentSseClient.connect(agent2.sessionId);

    await agent1.callTool('plan', {
      'action': 'update',
      'agent_key': key1,
      'goal': _testGoal,
      'current_task': _testTask,
    });

    final events = await sse.waitForEvents(1);
    sse.close();

    expect(
      events.isNotEmpty,
      isTrue,
      reason: 'Agent2 MUST receive SSE event for plan update',
    );
    final eventType = _extractEventType(events.first);
    expect(
      eventType,
      equals(_eventPlanUpdated),
      reason: 'Event type MUST be $_eventPlanUpdated',
    );
  });

  test('agent receives lock_released notification via SSE', () async {
    await _resetServer();
    final agent1 = _McpClient();
    final agent2 = _McpClient();
    await agent1.initSession();
    await agent2.initSession();

    final reg1 = _parseJson(
      await agent1.callTool('register', {'name': _agent1Name}),
    );
    await agent2.callTool('register', {'name': _agent2Name});
    final key1 = reg1['agent_key']! as String;

    // Acquire lock first
    await agent1.callTool('lock', {
      'action': 'acquire',
      'file_path': _testFilePath,
      'agent_key': key1,
      'reason': _testLockReason,
    });

    final sse = await _AgentSseClient.connect(agent2.sessionId);

    // Release the lock
    await agent1.callTool('lock', {
      'action': 'release',
      'file_path': _testFilePath,
      'agent_key': key1,
    });

    final events = await sse.waitForEvents(1);
    sse.close();

    expect(
      events.isNotEmpty,
      isTrue,
      reason: 'Agent2 MUST receive SSE event for lock release',
    );
    final eventType = _extractEventType(events.first);
    expect(
      eventType,
      equals(_eventLockReleased),
      reason: 'Event type MUST be $_eventLockReleased',
    );
  });

  test('agent notification has correct JSON-RPC payload structure', () async {
    await _resetServer();
    final agent1 = _McpClient();
    final agent2 = _McpClient();
    await agent1.initSession();
    await agent2.initSession();

    await agent1.callTool('register', {'name': _agent1Name});
    await agent2.callTool('register', {'name': _agent2Name});

    final sse = await _AgentSseClient.connect(agent2.sessionId);

    // Register agent3 to trigger a notification
    await agent1.callTool('register', {'name': _agent3Name});

    final events = await sse.waitForEvents(1);
    sse.close();

    expect(events.isNotEmpty, isTrue);

    final eventJson = _parseJson(events.first);

    // Verify JSON-RPC envelope
    expect(
      eventJson['jsonrpc'],
      equals(_jsonRpcVersion),
      reason: 'MUST have jsonrpc version',
    );
    expect(
      eventJson['method'],
      equals(_notificationMethod),
      reason: 'MUST have notifications/message method',
    );

    // Verify params structure
    final params = eventJson['params']! as Map<String, Object?>;
    expect(params['level'], equals(_levelInfo), reason: 'MUST have info level');

    // Verify data structure
    final data = params['data']! as Map<String, Object?>;
    expect(data.containsKey('event'), isTrue, reason: 'MUST contain event key');
    expect(
      data.containsKey('payload'),
      isTrue,
      reason: 'MUST contain payload key',
    );
    expect(
      data.containsKey('timestamp'),
      isTrue,
      reason: 'MUST contain timestamp key',
    );
    expect(
      data['event'],
      equals(_eventAgentRegistered),
      reason: 'Event type MUST be $_eventAgentRegistered',
    );
  });
}

// ============================================================
// Helper: parse JSON and extract event type
// ============================================================

Map<String, Object?> _parseJson(String text) =>
    jsonDecode(text) as Map<String, Object?>;

String? _extractEventType(String sseData) {
  final json = _parseJson(sseData);
  final params = json['params'] as Map<String, Object?>?;
  final data = params?['data'] as Map<String, Object?>?;
  return data?['event'] as String?;
}

// ============================================================
// Agent SSE Client — opens GET /mcp with agent session ID
// ============================================================

class _AgentSseClient {
  _AgentSseClient._();

  final _events = <String>[];
  var _consumed = 0;
  _SseReader? _reader;

  static Future<_AgentSseClient> connect(String sessionId) async {
    final client = _AgentSseClient._();
    client._reader = await _SseReader.open(
      '$_baseUrl$_mcpPath',
      sessionId,
      client._events,
    );
    await Future<void>.delayed(
      const Duration(milliseconds: _streamEstablishDelayMs),
    );
    return client;
  }

  Future<List<String>> waitForEvents(
    int count, {
    int timeoutMs = _defaultEventTimeoutMs,
  }) async {
    final start = DateTime.now().millisecondsSinceEpoch;
    while (DateTime.now().millisecondsSinceEpoch - start < timeoutMs) {
      if (_events.length - _consumed >= count) {
        final result = _events.sublist(_consumed);
        _consumed = _events.length;
        return result;
      }
      await Future<void>.delayed(
        const Duration(milliseconds: _eventPollDelayMs),
      );
    }
    final result = _events.sublist(_consumed);
    _consumed = _events.length;
    return result;
  }

  void close() {
    _reader?.abort();
  }
}

// ============================================================
// SSE stream reader
// ============================================================

class _SseReader {
  _SseReader._(this._controller);

  final JSObject _controller;
  static const _dataPrefix = 'data: ';

  static Future<_SseReader> open(
    String url,
    String sessionId,
    List<String> events,
  ) async {
    final controller = _createAbortController();
    final signal = controller['signal']!;

    final headers = JSObject()
      ..['Accept'] = _accept.toJS
      ..['mcp-session-id'] = sessionId.toJS;
    final options = JSObject()
      ..['method'] = 'GET'.toJS
      ..['headers'] = headers
      ..['signal'] = signal;

    unawaited(
      Future<void>(() async {
        try {
          final response = await _jsFetch(url.toJS, options).toDart;
          final ok = response['ok'] as JSBoolean?;
          if (ok == null || !ok.toDart) return;

          final body = response['body'];
          if (body == null || body.isUndefinedOrNull) {
            return;
          }

          final reader =
              ((body as JSObject)['getReader']! as JSFunction).callAsFunction(
                    body,
                  )!
                  as JSObject;
          final decoder = _createTextDecoder();
          var buffer = '';

          for (;;) {
            final chunk =
                await ((reader['read']! as JSFunction).callAsFunction(reader)!
                        as JSPromise<JSObject>)
                    .toDart;

            final done = chunk['done'] as JSBoolean?;
            if (done != null && done.toDart) break;

            final value = chunk['value'];
            if (value == null || value.isUndefinedOrNull) {
              continue;
            }

            final decoded =
                (decoder['decode']! as JSFunction).callAsFunction(
                      decoder,
                      value,
                      _streamOptions,
                    )!
                    as JSString;
            final buf = StringBuffer(buffer)..write(decoded.toDart);
            buffer = buf.toString();

            final lines = buffer.split('\n');
            buffer = lines.removeLast();
            for (final line in lines) {
              if (line.startsWith(_dataPrefix)) {
                final data = line.substring(_dataPrefix.length).trim();
                if (data.isNotEmpty) {
                  events.add(data);
                }
              }
            }
          }
        } on Object {
          // Stream aborted — expected on close()
        }
      }),
    );

    return _SseReader._(controller);
  }

  void abort() {
    (_controller['abort']! as JSFunction).callAsFunction(_controller);
  }
}

// ============================================================
// Shared HTTP / Node.js helpers
// ============================================================

@JS('globalThis.fetch')
external JSPromise<JSObject> _jsFetch(JSString url, [JSObject? options]);

@JS('globalThis.AbortController')
external JSFunction get _abortControllerCtor;

JSObject _createAbortController() =>
    _abortControllerCtor.callAsConstructor<JSObject>();

@JS('globalThis.TextDecoder')
external JSFunction get _textDecoderCtor;

JSObject _createTextDecoder() => _textDecoderCtor.callAsConstructor<JSObject>();

final JSObject _streamOptions = JSObject()..['stream'] = true.toJS;

String? _getResponseHeader(JSObject response, String name) {
  final headers = response['headers'] as JSObject?;
  if (headers == null) return null;
  final getFn = headers['get'] as JSFunction?;
  final value = getFn?.callAsFunction(headers, name.toJS);
  if (value == null || value.isUndefinedOrNull) return null;
  return (value as JSString).toDart;
}

// ============================================================
// MCP Client with exposed session ID
// ============================================================

class _McpClient {
  String _sessionId = '';
  var _nextId = 1;

  String get sessionId {
    if (_sessionId.isEmpty) {
      throw StateError('Session not initialized');
    }
    return _sessionId;
  }

  Future<void> initSession() async {
    final initResult = await _request('initialize', {
      'protocolVersion': _mcpProtocolVersion,
      'capabilities': <String, Object?>{},
      'clientInfo': {'name': 'agent-streaming-e2e', 'version': '1.0.0'},
    });
    if (_sessionId.isEmpty) {
      throw StateError('No session ID after init: $initResult');
    }
    await _postMcp(
      jsonEncode({
        'jsonrpc': _jsonRpcVersion,
        'method': 'notifications/initialized',
        'params': <String, Object?>{},
      }),
    );
  }

  Future<String> callTool(String name, Map<String, Object?> args) async {
    final result = await _request('tools/call', {
      'name': name,
      'arguments': args,
    });
    final content = (result['content']! as List).first as Map<String, Object?>;
    return content['text']! as String;
  }

  Future<Map<String, Object?>> _request(
    String method,
    Map<String, Object?> params,
  ) async {
    final id = _nextId++;
    final body = jsonEncode({
      'jsonrpc': _jsonRpcVersion,
      'id': id,
      'method': method,
      'params': params,
    });
    final response = await _postMcp(body);
    final text = await _responseText(response);
    final json = _parseMcpResponse(text);
    if (json.containsKey('error')) {
      final error = json['error']! as Map<String, Object?>;
      final message = error['message'] as String? ?? 'Error';
      return <String, Object?>{
        'isError': true,
        'content': <Object>[
          <String, Object?>{'type': 'text', 'text': message},
        ],
      };
    }
    return json['result']! as Map<String, Object?>;
  }

  Future<JSObject> _postMcp(String body) async {
    final headers = JSObject()
      ..['Content-Type'] = 'application/json'.toJS
      ..['Accept'] = _accept.toJS;
    if (_sessionId.isNotEmpty) {
      headers['mcp-session-id'] = _sessionId.toJS;
    }
    final options = JSObject()
      ..['method'] = 'POST'.toJS
      ..['headers'] = headers
      ..['body'] = body.toJS;
    final response = await _jsFetch('$_baseUrl$_mcpPath'.toJS, options).toDart;
    final sid = _getResponseHeader(response, 'mcp-session-id');
    if (sid != null) _sessionId = sid;
    return response;
  }

  Future<String> _responseText(JSObject response) async {
    final text =
        await ((response['text'] as JSFunction?)?.callAsFunction(response)
                as JSPromise<JSString>?)
            ?.toDart;
    return text?.toDart ?? '';
  }

  Map<String, Object?> _parseMcpResponse(String text) {
    if (text.trimLeft().startsWith('{')) {
      return jsonDecode(text) as Map<String, Object?>;
    }
    for (final line in text.split('\n')) {
      if (line.startsWith('data: ')) {
        try {
          return jsonDecode(line.substring(6)) as Map<String, Object?>;
        } on Object {
          continue;
        }
      }
    }
    throw StateError('Could not parse: $text');
  }
}

// ============================================================
// Server lifecycle helpers
// ============================================================

JSObject _spawnServer() {
  final childProcess = requireModule('child_process') as JSObject;
  final spawnFn = childProcess['spawn']! as JSFunction;
  return spawnFn.callAsFunction(
        null,
        'node'.toJS,
        <String>[serverBinary].jsify(),
        <String, Object?>{
          'stdio': ['pipe', 'pipe', 'inherit'],
        }.jsify(),
      )!
      as JSObject;
}

void _killProcess(JSObject process) {
  (process['kill']! as JSFunction).callAsFunction(process);
}

Future<void> _waitForServer() async {
  for (var i = 0; i < _maxServerPollAttempts; i++) {
    try {
      final r = await _jsFetch('$_baseUrl/admin/status'.toJS).toDart;
      final ok = r['ok'] as JSBoolean?;
      if (ok != null && ok.toDart) return;
    } on Object {
      // Not ready yet
    }
    if (i == _maxServerPollAttempts - 1) {
      throw StateError('Server failed to start');
    }
    await Future<void>.delayed(
      const Duration(milliseconds: _serverPollDelayMs),
    );
  }
}

Future<void> _resetServer() async {
  final options = JSObject()
    ..['method'] = 'POST'.toJS
    ..['headers'] = (JSObject()..['Content-Type'] = 'application/json'.toJS);
  final r = await _jsFetch('$_baseUrl/admin/reset'.toJS, options).toDart;
  final ok = r['ok'] as JSBoolean?;
  if (ok == null || !ok.toDart) {
    throw StateError('Failed to reset server');
  }
}

void _deleteDbFiles() {
  final fs = requireModule('fs') as JSObject;
  final unlinkSync = fs['unlinkSync']! as JSFunction;
  final existsSync = fs['existsSync']! as JSFunction;

  for (final file in _dbFiles) {
    final path = '$_dbDir/$file';
    final exists =
        (existsSync.callAsFunction(fs, path.toJS) as JSBoolean?)?.toDart ??
        false;
    if (exists) {
      unlinkSync.callAsFunction(fs, path.toJS);
    }
  }
}

File: too-many-cooks/test/agent_streaming_e2e_test.ts
/// E2E agent streaming test — spawn MCP server, open SSE stream
/// on /mcp for an AGENT session, trigger state changes from
/// another agent, ASSERT that notifications arrive over the
/// agent's SSE stream.
///
/// This PROVES that agents receive streamed notifications over
/// their Streamable HTTP SSE connection.

import { describe, it, before, after } from "node:test";
import assert from "node:assert";

import { spawn, type ChildProcess } from "node:child_process";
import fs from "node:fs";

import { SERVER_BINARY, SERVER_NODE_ARGS } from "../lib/src/config.js";

// ============================================================
// Named Constants
// ============================================================

const TEST_PORT = 4046;
const BASE_URL = `http://localhost:${String(TEST_PORT)}` as const;
const MCP_PATH = "/mcp" as const;
const ACCEPT = "application/json, text/event-stream" as const;
const MCP_PROTOCOL_VERSION = "2025-03-26" as const;
const DEFAULT_EVENT_TIMEOUT_MS = 3000;
const STREAM_ESTABLISH_DELAY_MS = 200;
const SERVER_POLL_DELAY_MS = 200;
const EVENT_POLL_DELAY_MS = 50;
const MAX_SERVER_POLL_ATTEMPTS = 30;

const AGENT_1_NAME = "agent-stream-1" as const;
const AGENT_2_NAME = "agent-stream-2" as const;
const AGENT_3_NAME = "agent-stream-3" as const;

const TEST_FILE_PATH = "/agent-stream/test.dart" as const;
const TEST_LOCK_REASON = "agent streaming e2e" as const;
const TEST_MESSAGE_CONTENT = "hello from agent1" as const;
const TEST_GOAL = "Agent streaming goal" as const;
const TEST_TASK = "Testing agent streaming" as const;

const EVENT_AGENT_REGISTERED = "agent_registered" as const;
const EVENT_LOCK_ACQUIRED = "lock_acquired" as const;
const EVENT_LOCK_RELEASED = "lock_released" as const;
const EVENT_MESSAGE_SENT = "message_sent" as const;
const EVENT_PLAN_UPDATED = "plan_updated" as const;

const JSON_RPC_VERSION = "2.0" as const;
const NOTIFICATION_METHOD = "notifications/message" as const;
const LEVEL_INFO = "info" as const;

const DB_DIR = ".too_many_cooks" as const;
const DB_FILES = ["data.db", "data.db-wal", "data.db-shm"] as const;

// ============================================================
// Helper: sleep
// ============================================================

const sleep = async (ms: number): Promise<void> =>
  new Promise((resolve) => setTimeout(resolve, ms));

// ============================================================
// Helper: parse JSON and extract event type
// ============================================================

const parseJson = (text: string): Record<string, unknown> =>
  JSON.parse(text) as Record<string, unknown>;

const extractEventType = (sseData: string): string | undefined => {
  const json = parseJson(sseData);
  const params = json.params as Record<string, unknown> | undefined;
  const data = params?.data as Record<string, unknown> | undefined;
  return data?.event as string | undefined;
};

// ============================================================
// Agent SSE Client — opens GET /mcp with agent session ID
// ============================================================

class AgentSseClient {
  private readonly events: string[] = [];
  private consumed = 0;
  private controller: AbortController | undefined;

  static async connect(sessionId: string): Promise<AgentSseClient> {
    const client = new AgentSseClient();
    client.controller = new AbortController();

    const headers: Record<string, string> = {
      Accept: ACCEPT,
      "mcp-session-id": sessionId,
    };

    void (async () => {
      try {
        const response = await fetch(`${BASE_URL}${MCP_PATH}`, {
          method: "GET",
          headers,
          signal: client.controller!.signal,
        });
        if (!response.ok || response.body === null) {return;}

        const reader = response.body.getReader();
        const decoder = new TextDecoder();
        let buffer = "";

        for (;;) {
          const { done, value } = await reader.read();
          if (done) {break;}

          buffer += decoder.decode(value, { stream: true });
          const lines = buffer.split("\n");
          buffer = lines.pop()!;
          for (const line of lines) {
            if (line.startsWith("data: ")) {
              const data = line.substring(6).trim();
              if (data.length > 0) {
                client.events.push(data);
              }
            }
          }
        }
      } catch {
        // Stream aborted — expected on close()
      }
    })();

    await sleep(STREAM_ESTABLISH_DELAY_MS);
    return client;
  }

  async waitForEvents(
    count: number,
    timeoutMs: number = DEFAULT_EVENT_TIMEOUT_MS,
  ): Promise<string[]> {
    const start = Date.now();
    while (Date.now() - start < timeoutMs) {
      if (this.events.length - this.consumed >= count) {
        const result = this.events.slice(this.consumed);
        this.consumed = this.events.length;
        return result;
      }
      await sleep(EVENT_POLL_DELAY_MS);
    }
    const result = this.events.slice(this.consumed);
    this.consumed = this.events.length;
    return result;
  }

  close(): void {
    this.controller?.abort();
  }
}

// ============================================================
// MCP Client with exposed session ID
// ============================================================

class McpClient {
  private _sessionId = "";
  private nextId = 1;

  get sessionId(): string {
    if (this._sessionId.length === 0) {
      throw new Error("Session not initialized");
    }
    return this._sessionId;
  }

  async initSession(): Promise<void> {
    const initResult = await this.request("initialize", {
      protocolVersion: MCP_PROTOCOL_VERSION,
      capabilities: {},
      clientInfo: { name: "agent-streaming-e2e", version: "1.0.0" },
    });
    if (this._sessionId.length === 0) {
      throw new Error(
        `No session ID after init: ${JSON.stringify(initResult)}`,
      );
    }
    await this.postMcp(
      JSON.stringify({
        jsonrpc: JSON_RPC_VERSION,
        method: "notifications/initialized",
        params: {},
      }),
    );
  }

  async callTool(name: string, args: Record<string, unknown>): Promise<string> {
    const result = await this.request("tools/call", {
      name,
      arguments: args,
    });
    const content = (result.content as Array<Record<string, unknown>>)[0];
    return content.text as string;
  }

  private async request(
    method: string,
    params: Record<string, unknown>,
  ): Promise<Record<string, unknown>> {
    const id = this.nextId++;
    const body = JSON.stringify({
      jsonrpc: JSON_RPC_VERSION,
      id,
      method,
      params,
    });
    const response = await this.postMcp(body);
    const text = await response.text();
    const json = this.parseMcpResponse(text);
    if ("error" in json) {
      const err = json.error as Record<string, unknown>;
      const message = (err.message as string | undefined) ?? "Error";
      return {
        isError: true,
        content: [{ type: "text", text: message }],
      };
    }
    return json.result as Record<string, unknown>;
  }

  private async postMcp(body: string): Promise<Response> {
    const headers: Record<string, string> = {
      "Content-Type": "application/json",
      Accept: ACCEPT,
    };
    if (this._sessionId.length > 0) {
      headers["mcp-session-id"] = this._sessionId;
    }
    const response = await fetch(`${BASE_URL}${MCP_PATH}`, {
      method: "POST",
      headers,
      body,
    });
    const sid = response.headers.get("mcp-session-id");
    if (sid !== null) {this._sessionId = sid;}
    return response;
  }

  private parseMcpResponse(text: string): Record<string, unknown> {
    if (text.trimStart().startsWith("{")) {
      return JSON.parse(text) as Record<string, unknown>;
    }
    for (const line of text.split("\n")) {
      if (line.startsWith("data: ")) {
        try {
          return JSON.parse(line.substring(6)) as Record<string, unknown>;
        } catch {
          continue;
        }
      }
    }
    throw new Error(`Could not parse: ${text}`);
  }
}

// ============================================================
// Server lifecycle helpers
// ============================================================

const spawnServer = (): ChildProcess =>
  spawn("node", [...SERVER_NODE_ARGS, SERVER_BINARY], {
    stdio: ["pipe", "pipe", "inherit"],
    env: { ...process.env, TMC_PORT: String(TEST_PORT) },
  });

const waitForServer = async (): Promise<void> => {
  for (let i = 0; i < MAX_SERVER_POLL_ATTEMPTS; i++) {
    try {
      const r = await fetch(`${BASE_URL}/admin/status`);
      if (r.ok) {return;}
    } catch {
      // Not ready yet
    }
    if (i === MAX_SERVER_POLL_ATTEMPTS - 1) {
      throw new Error("Server failed to start");
    }
    await sleep(SERVER_POLL_DELAY_MS);
  }
};

const resetServer = async (): Promise<void> => {
  const r = await fetch(`${BASE_URL}/admin/reset`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
  });
  if (!r.ok) {
    throw new Error("Failed to reset server");
  }
};

const deleteDbFiles = (): void => {
  for (const file of DB_FILES) {
    const path = `${DB_DIR}/${file}`;
    try {
      if (fs.existsSync(path)) {
        fs.unlinkSync(path);
      }
    } catch {
      // ignore
    }
  }
};

// ============================================================
// Tests
// ============================================================

describe("agent_streaming_e2e_test", () => {
  let serverProcess: ChildProcess;

  before(async () => {
    deleteDbFiles();
    serverProcess = spawnServer();
    await waitForServer();
  });

  after(() => {
    serverProcess.kill();
    deleteDbFiles();
  });

  it("agent receives message_sent notification via SSE", async () => {
    await resetServer();
    const agent1 = new McpClient();
    const agent2 = new McpClient();
    await agent1.initSession();
    await agent2.initSession();

    const reg1 = parseJson(
      await agent1.callTool("register", { name: AGENT_1_NAME }),
    );
    const reg2 = parseJson(
      await agent2.callTool("register", { name: AGENT_2_NAME }),
    );
    const key1 = reg1.agent_key as string;

    const sse = await AgentSseClient.connect(agent2.sessionId);

    await agent1.callTool("message", {
      action: "send",
      agent_key: key1,
      to_agent: reg2.agent_name as string,
      content: TEST_MESSAGE_CONTENT,
    });

    const events = await sse.waitForEvents(1);
    sse.close();

    assert.strictEqual(events.length > 0, true);
    const eventType = extractEventType(events[0]);
    assert.strictEqual(eventType, EVENT_MESSAGE_SENT);
  });

  it("agent receives lock_acquired notification via SSE", async () => {
    await resetServer();
    const agent1 = new McpClient();
    const agent2 = new McpClient();
    await agent1.initSession();
    await agent2.initSession();

    const reg1 = parseJson(
      await agent1.callTool("register", { name: AGENT_1_NAME }),
    );
    await agent2.callTool("register", { name: AGENT_2_NAME });
    const key1 = reg1.agent_key as string;

    const sse = await AgentSseClient.connect(agent2.sessionId);

    await agent1.callTool("lock", {
      action: "acquire",
      file_path: TEST_FILE_PATH,
      agent_key: key1,
      reason: TEST_LOCK_REASON,
    });

    const events = await sse.waitForEvents(1);
    sse.close();

    assert.strictEqual(events.length > 0, true);
    const eventType = extractEventType(events[0]);
    assert.strictEqual(eventType, EVENT_LOCK_ACQUIRED);
  });

  it("agent receives agent_registered notification via SSE", async () => {
    await resetServer();
    const agent1 = new McpClient();
    const agent2 = new McpClient();
    await agent1.initSession();
    await agent2.initSession();

    await agent1.callTool("register", { name: AGENT_1_NAME });
    await agent2.callTool("register", { name: AGENT_2_NAME });

    const sse = await AgentSseClient.connect(agent2.sessionId);

    // Register a third agent — agent2 should get notified
    await agent1.callTool("register", { name: AGENT_3_NAME });

    const events = await sse.waitForEvents(1);
    sse.close();

    assert.strictEqual(events.length > 0, true);
    const eventType = extractEventType(events[0]);
    assert.strictEqual(eventType, EVENT_AGENT_REGISTERED);
  });

  it("agent receives plan_updated notification via SSE", async () => {
    await resetServer();
    const agent1 = new McpClient();
    const agent2 = new McpClient();
    await agent1.initSession();
    await agent2.initSession();

    const reg1 = parseJson(
      await agent1.callTool("register", { name: AGENT_1_NAME }),
    );
    await agent2.callTool("register", { name: AGENT_2_NAME });
    const key1 = reg1.agent_key as string;

    const sse = await AgentSseClient.connect(agent2.sessionId);

    await agent1.callTool("plan", {
      action: "update",
      agent_key: key1,
      goal: TEST_GOAL,
      current_task: TEST_TASK,
    });

    const events = await sse.waitForEvents(1);
    sse.close();

    assert.strictEqual(events.length > 0, true);
    const eventType = extractEventType(events[0]);
    assert.strictEqual(eventType, EVENT_PLAN_UPDATED);
  });

  it("agent receives lock_released notification via SSE", async () => {
    await resetServer();
    const agent1 = new McpClient();
    const agent2 = new McpClient();
    await agent1.initSession();
    await agent2.initSession();

    const reg1 = parseJson(
      await agent1.callTool("register", { name: AGENT_1_NAME }),
    );
    await agent2.callTool("register", { name: AGENT_2_NAME });
    const key1 = reg1.agent_key as string;

    // Acquire lock first
    await agent1.callTool("lock", {
      action: "acquire",
      file_path: TEST_FILE_PATH,
      agent_key: key1,
      reason: TEST_LOCK_REASON,
    });

    const sse = await AgentSseClient.connect(agent2.sessionId);

    // Release the lock
    await agent1.callTool("lock", {
      action: "release",
      file_path: TEST_FILE_PATH,
      agent_key: key1,
    });

    const events = await sse.waitForEvents(1);
    sse.close();

    assert.strictEqual(events.length > 0, true);
    const eventType = extractEventType(events[0]);
    assert.strictEqual(eventType, EVENT_LOCK_RELEASED);
  });

  it("agent notification has correct JSON-RPC payload structure", async () => {
    await resetServer();
    const agent1 = new McpClient();
    const agent2 = new McpClient();
    await agent1.initSession();
    await agent2.initSession();

    await agent1.callTool("register", { name: AGENT_1_NAME });
    await agent2.callTool("register", { name: AGENT_2_NAME });

    const sse = await AgentSseClient.connect(agent2.sessionId);

    // Register agent3 to trigger a notification
    await agent1.callTool("register", { name: AGENT_3_NAME });

    const events = await sse.waitForEvents(1);
    sse.close();

    assert.strictEqual(events.length > 0, true);

    const eventJson = parseJson(events[0]);

    // Verify JSON-RPC envelope
    assert.strictEqual(eventJson.jsonrpc, JSON_RPC_VERSION);
    assert.strictEqual(eventJson.method, NOTIFICATION_METHOD);

    // Verify params structure
    const params = eventJson.params as Record<string, unknown>;
    assert.strictEqual(params.level, LEVEL_INFO);

    // Verify data structure
    const data = params.data as Record<string, unknown>;
    assert.strictEqual("event" in data, true);
    assert.strictEqual("payload" in data, true);
    assert.strictEqual("timestamp" in data, true);
    assert.strictEqual(data.event, EVENT_AGENT_REGISTERED);
  });
});
